src/librustc_typeck/check/writeback.rs

```diff
@@ -190,9 +190,23 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         if let hir::ExprKind::Index(ref base, ref index) = e.node {
             let mut tables = self.fcx.tables.borrow_mut();
 
-            // All valid indexing looks like this; might encounter non-valid indexes at this point
-            if let ty::Ref(_, base_ty, _) = tables.expr_ty_adjusted(&base).sty {
-                let index_ty = tables.expr_ty_adjusted(&index);
+            // All valid indexing looks like this; might encounter non-valid indexes at this point.
+            let base_ty = tables.expr_ty_adjusted_opt(&base).map(|t| &t.sty);
+            if base_ty.is_none() {
+                self.tcx().sess.delay_span_bug(e.span, &format!("bad base: `{:?}`", base));
+                return;
+            }
+            if let Some(ty::Ref(_, base_ty, _)) = base_ty {
+                let index_ty = match tables.expr_ty_adjusted_opt(&index) {
```

Comment:
```suggestion
                let index_ty = tables.expr_ty_adjusted_opt(&index).unwrap_or_else(|| {
```

Reply:
Originally didn't go down that route because I was thinking about `self` borrowing, but `self` is not involved here :)